scipy/_lib/_array_api.py

```diff
@@ -195,33 +196,78 @@ def is_torch(xp):
     return xp.__name__ == 'scipy._lib.array_api_compat.array_api_compat.torch'
 
 
-def assert_equal(actual, desired, err_msg='', xp=None):
+def _strict_check(actual, desired, check_xp=True, check_dtype=True, check_shape=True):
+    try:
+        actual.dtype
+        desired.dtype
+    except AttributeError:
+        raise ValueError("Inputs should be arrays with a dtype attribute, "
+                         "python scalars and arrays are not accepted.")
+    if check_xp:
+        _assert_matching_namespace(actual, desired)
+    if check_dtype:
+        assert_(actual.dtype == desired.dtype,
+                f"Desired dtype: {desired.dtype}, actual dtype: {actual.dtype}")
+    if check_shape:
+        msg = (f"Array shapes are not equal: actual shape = {actual.shape}, "
+               f"desired shape = {desired.shape}")
+        assert_(actual.shape == desired.shape, msg)
+
+
+def _assert_matching_namespace(actual, desired):
+    desired_space = array_namespace(desired)
+    if isinstance(actual, tuple):
+        for arr in actual:
+            arr_space = array_namespace(arr)
+            assert arr_space == desired_space
+            if is_numpy(arr_space):
+                _check_scalar(actual, desired)
+    else:
+        actual_space = array_namespace(actual)
+        assert_(actual_space == desired_space,
+                "Namespaces do not match: "
+                f"{actual_space.__name__} and {desired_space.__name__}")
+        if is_numpy(actual_space):
```

Comment:
It looks like CuPy makes a distinction between scalars and 0d array, too. However, it could be argued that testing for scalars vs arrays is not as important for CuPy as it is for NumPy because it is much harder to get a scalar out of CuPy:
```python3
import cupyx
cupyx.scipy.special.ndtr(0)[()]
```
That is, in CuPy, functions return a 0d array even when given a scalar, and using `[()]` doesn't extract a scalar, so it's much harder to get inconsistencies accidentally.

Reply:
It does only apply to NumPy. I think that's probably fine, but if you want to check for other libraries as well then we could.

Reply:
> It looks like CuPy makes a distinction between scalars and 0d array, too.

There are no scalars in CuPy, Pytorch or any other relevant library. They all use 0-D arrays and that works just fine. We'd remove scalars from NumPy if we could, but that ship has sailed.